crates/composition/src/compose_supergraph/enums.rs

```diff
@@ -0,0 +1,134 @@
+    };
+    let in_input_type_fields = || {
+        subgraphs
+            .iter_fields()
+            .filter(|field| field.parent_definition().kind() == DefinitionKind::InputObject)
+            .any(|field| field.r#type().type_name() == enum_name)
+    };
+    in_field_arguments() || in_input_type_fields()
+}
+
+/// Returns whether the enum is returned by a field anywhere.
+fn enum_is_used_in_return_position(enum_name: StringId, subgraphs: &Subgraphs) -> bool {
+    subgraphs
+        .iter_fields()
+        .filter(|field| {
+            matches!(
+                field.parent_definition().kind(),
+                DefinitionKind::Object | DefinitionKind::Interface
+            )
+        })
+        .any(|field| field.r#type().type_name() == enum_name)
+}
+
+fn merge_intersection(
+    first: &DefinitionWalker<'_>,
+    definitions: &[DefinitionWalker<'_>],
+    ctx: &mut Context<'_>,
+) {
+    let mut intersection: Vec<StringId> = first.enum_values().collect();
+    let mut scratch = HashSet::new();
+
+    for definition in definitions {
+        scratch.clear();
+        scratch.extend(definition.enum_values());
+        intersection.retain(|elem| scratch.contains(elem));
+    }
+
+    if intersection.is_empty() {
+        ctx.diagnostics.push_fatal(format!(
+            "Values for enum {} are empty (intersection)",
+            first.name_str(),
+        ));
+    }
+
+    ctx.supergraph
+        .insert_definition(first.name(), DefinitionKind::Enum);
+
+    for value in intersection {
+        ctx.supergraph.insert_enum_value(first.name(), value)
+    }
+}
+
+fn merge_union(
+    first: &DefinitionWalker<'_>,
+    definitions: &[DefinitionWalker<'_>],
+    ctx: &mut Context<'_>,
+) {
+    ctx.supergraph
+        .insert_definition(first.name(), DefinitionKind::Enum);
+
+    for value in definitions.iter().flat_map(|def| def.enum_values()) {
+        ctx.supergraph.insert_enum_value(first.name(), value)
+    }
+}
+
+fn merge_exactly_matching(
+    first: &DefinitionWalker<'_>,
+    definitions: &[DefinitionWalker<'_>],
+    ctx: &mut Context<'_>,
+) {
+    let expected: Vec<_> = first.enum_values().collect();
+
+    for definition in definitions {
+        if !is_slice_match(&expected, definition.enum_values()) {
+            ctx.diagnostics.push_fatal(format!(
+                "The enum {} should match exactly in all subgraphs, but it does not",
+                first.name_str()
+            ));
+            return;
+        }
+    }
+
+    ctx.supergraph
+        .insert_definition(first.name(), DefinitionKind::Enum);
+
+    for value in expected {
+        ctx.supergraph.insert_enum_value(first.name(), value)
+    }
+}
+
+fn is_slice_match<T: PartialEq>(slice: &[T], iterator: impl Iterator<Item = T>) -> bool {
+    let mut idx = 0;
+
+    for item in iterator {
+        if slice[idx] != item {
+            return false;
+        }
+
+        idx += 1;
+    }
+
+    idx == slice.len()
+}
```

Comment:
If someone has something shorter / easier to read with std or itertools (haven't tried zip_longest(), afraid of the item enum), I'll take it gladly.